app/workers/work_packages/update_progress_job.rb

```diff
@@ -0,0 +1,257 @@
+    execute(<<~SQL.squish)
+      UPDATE temp_wp_progress_values
+      SET done_ratio = statuses.default_done_ratio
+      FROM statuses
+      WHERE temp_wp_progress_values.status_id = statuses.id
+    SQL
+  end
+
+  def update_totals
+    execute(<<~SQL)
+      UPDATE temp_wp_progress_values
+      SET total_work = totals.total_work,
+          total_remaining_work = totals.total_remaining_work,
+          total_p_complete = CASE
+            WHEN totals.total_work = 0 THEN NULL
+            ELSE (1 - (totals.total_remaining_work / totals.total_work)) * 100
+          END
+      FROM (
+        SELECT wp_tree.ancestor_id AS id,
+               SUM(estimated_hours) AS total_work,
+               SUM(remaining_hours) AS total_remaining_work
+        FROM work_package_hierarchies wp_tree
+          LEFT JOIN temp_wp_progress_values wp_progress ON wp_tree.descendant_id = wp_progress.id
+        GROUP BY wp_tree.ancestor_id
+      ) totals
+      WHERE temp_wp_progress_values.id = totals.id
+    SQL
+  end
+
+  def copy_progress_values_to_work_packages_and_update_journals
+    updated_work_package_ids = copy_progress_values_to_work_packages
+    create_journals_for_updated_work_packages(updated_work_package_ids)
+  end
+
+  def copy_progress_values_to_work_packages
+    results = execute(<<~SQL)
+      UPDATE work_packages
+      SET estimated_hours = temp_wp_progress_values.estimated_hours,
+          remaining_hours = temp_wp_progress_values.remaining_hours,
+          done_ratio = temp_wp_progress_values.done_ratio,
+          derived_estimated_hours = temp_wp_progress_values.total_work,
+          derived_remaining_hours = temp_wp_progress_values.total_remaining_work,
+          derived_done_ratio = temp_wp_progress_values.total_p_complete,
+          lock_version = lock_version + 1,
+          updated_at = NOW()
+      FROM temp_wp_progress_values
+      WHERE work_packages.id = temp_wp_progress_values.id
+        AND (
+          work_packages.estimated_hours IS DISTINCT FROM temp_wp_progress_values.estimated_hours
+          OR work_packages.remaining_hours IS DISTINCT FROM temp_wp_progress_values.remaining_hours
+          OR work_packages.done_ratio IS DISTINCT FROM temp_wp_progress_values.done_ratio
+          OR work_packages.derived_estimated_hours IS DISTINCT FROM temp_wp_progress_values.total_work
+          OR work_packages.derived_remaining_hours IS DISTINCT FROM temp_wp_progress_values.total_remaining_work
+          OR work_packages.derived_done_ratio IS DISTINCT FROM temp_wp_progress_values.total_p_complete
+        )
+      RETURNING work_packages.id
+    SQL
+    results.column_values(0)
+  end
+
+  def create_journals_for_updated_work_packages(updated_work_package_ids)
+    WorkPackage.where(id: updated_work_package_ids).find_each do |work_package|
+      Journals::CreateService.new(work_package, system_user)
+        .call(cause: { type: "system_update", feature: "progress_calculation_changed" })
+    end
```

Comment:
Here, it could happen (don't know how likely it is), that the job runs, updates the values in the `work _packages` table but hasn't completed the journal creation yet, that a user comes along and updates the work package. The updates to the work/done_ratio would be attributed to the user.

The larger the dataset, the more likely something like this is to happen. That being said, it might be something we are willing to accept. 

A way around that might be to do the updating of the work packages one at a time (potentially in combination with aquiring [advisory locks as done by the services](https://github.com/opf/openproject/blob/dev/app/services/shared/service_context.rb#L44)). Admittedly, that would be slower to execute.

Reply:
That's a valid point. But why would the modification be attributed to the user? The change is in a transaction (one created by the job) and the transaction will only be commited at the end of the job, so if a user changes something, this user will not see the not-yet-commited changes of the job, right?

Maybe I'm wrong and the job does not crete a transaction on its own.
I just checked and it seems the job does not create any transaction. I need to change that then.

Not completely sure of how this could be fully prevented. I'll add it as a TODO.